File: src/ast/Identifier.php
<?php

/**
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @license http://www.gnu.org/licenses/ GNU General Public License
 * @author  Sven Strittmatter <ich@weltraumschaf.de>
 * @package ast
 */

namespace de\weltraumschaf\ebnf\ast;

/**
 * @see Node
 */
require_once __DIR__ . DIRECTORY_SEPARATOR . 'Node.php';
/**
 * @see Type
 */
require_once __DIR__ . DIRECTORY_SEPARATOR . 'Type.php';

use de\weltraumschaf\ebnf\visitor\Visitor as Visitor;

/**
 * Identifier node.
 *
 * Has no subnodes.
 *
 * @package ast
 */
class Identifier implements Node {

    /**
     * The literal string.
     *
     * @var string
     */
    public $value = "";

    /**
     * Returns the name of a node.
     *
     * @return string
     */
    public function getNodeName() {
        return Type::IDENTIFIER;
    }

    /**
     * Defines method to accept {@link Visitors}.
     *
     * Imlements {@link http://en.wikipedia.org/wiki/Visitor_pattern Visitor Pattern}.
     *
     * @param Visitor $visitor Object which visits te node.
     *
     * @return void
     */
    public function accept(Visitor $visitor) {
        $visitor->beforeVisit($this);
        $visitor->visit($this);
        $visitor->afterVisit($this);
    }

    protected function probeEquivalenceInternal(Node $other, Notification $result) {
        if ( ! $other instanceof Identifier) {
            $result->error("Probed node types mismatch: '%s' != '%s'!", get_class($this), get_class($other));
            return;
        }

        if ($this->value !== $other->value) {
            $this->error("Identifier value mismatch: %s != %s!", $this->value, $other->value);
        }
    }
}

File: java/src/main/java/de/weltraumschaf/ebnf/ast/Identifier.java
package de.weltraumschaf.ebnf.ast;

import de.weltraumschaf.ebnf.visitor.Visitor;

/**
 * Identifier node.
 *
 * Has no sub nodes.
 *
 * @author Sven Strittmatter <weltraumschaf@googlemail.com>
 */
public class Identifier extends AbstractNode {

    /**
     * The literal string.
     *
     * @var
     */
    public String value = "";

    /**
     * Initializes object with empty value and parent node.
     *
     * @param Node The parent node.
     */
    public Identifier(Node parent) {
        super(parent);
    }

    /**
     * Returns the name of a node.
     *
     * @return
     */
    @Override
    public String getNodeName() {
        return NodeType.IDENTIFIER.toString();
    }

    /**
     * Defines method to accept {@link Visitors}.
     *
     * Implements {@link http://en.wikipedia.org/wiki/Visitor_pattern Visitor Pattern}.
     *
     * @param Visitor Object which visits te node.
     *
     * @return
     */
    @Override
    public void accept(Visitor visitor) {
        visitor.beforeVisit(this);
        visitor.visit(this);
        visitor.afterVisit(this);
    }

    /**
     * Probes equivalence of itself against an other node and collects all
     * errors in the passed {@link Notification} object.
     *
     * @param Node         Node to compare against.
     * @param Notification Object which collects all equivalence violations.
     *
     * @return
     */
    @Override
    public void probeEquivalence(Node other, Notification result) {
        try {
            Identifier ident = (Identifier) other;

            if (!value.equals(ident.value)) {
                result.error("Identifier value mismatch: '%s' != '%s'!", value, ident.value);
            }
        } catch (ClassCastException ex) {
            result.error("Probed node types mismatch: '%s' != '%s'!", getClass(), other.getClass());
        }
    }

    /**
     * Always returns 1.
     *
     * @return
     */
    @Override
    public int depth() {
        return 1;
    }

}
